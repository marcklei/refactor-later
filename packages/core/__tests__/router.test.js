const { applyParams, getRoute, mergeRoutes, prepareRoutes } = require('../lib/router')

const someRoutes = {
  POST: [{
    path: '/group/:id',
    handler: (req, res) => {
      res.write('Post on Group')
      res.end()
    }
  }],
  GET: [{
    path: '/',
    handler: (req, res) => {
      res.write('Get on Root')
      res.end()
    }
  }]
}

describe('getRoute ...', () => {
  test('returns undefined when no method is matched', () => {
    const result = getRoute({ POST: [] }, { method: 'GET' })
    expect(result).toBe(undefined)
  })

  test('returns undefined when no route is matched', () => {
    const match = jest.fn(() => false)
    const result = getRoute({ GET: [{ match }] }, { method: 'GET' })

    expect(match.mock.calls.length).toBe(1)
    expect(result).toBe(undefined)
  })

  test('returns the route object a route is matched', () => {
    const match = jest.fn(() => true)
    const route = { match, path: '/' }
    const result = getRoute({ GET: [route] }, { method: 'GET' })

    expect(match.mock.calls.length).toBe(1)
    expect(result.path).toBe('/')
  })
})

describe('mergeRoutes ...', () => {
  test('returns a merged routes object', () => {
    const moreRoutes = {
      POST: [{
        path: '/user/:id',
        handler: (req, res) => {
          res.write('Post on Root')
          res.end()
        }
      }],
      GET: [{
        path: '/testroute',
        handler: (req, res) => {
          res.write('Get on Testroute')
          res.end()
        }
      }]
    }

    const mergedRoutes = mergeRoutes(someRoutes, moreRoutes)

    expect(mergedRoutes).toMatchSnapshot()
  })
})

describe('prepareRoutes ...', () => {
  test('adds a regex matcher to each route', () => {
    const preparedRoutes = prepareRoutes(someRoutes)

    expect(typeof preparedRoutes['GET'][0].match).toBe('function')
    expect(typeof preparedRoutes['POST'][0].match).toBe('function')

    expect(typeof preparedRoutes['GET'][0].match('/')).toBeDefined()
    expect(preparedRoutes['POST'][0].match('/group/:id')).toBeDefined()
    expect(preparedRoutes['POST'][0].match('/')).toBeNull()
  })
})

describe('applyParams ...', () => {
  test('applys the params and query to the req', () => {
    const req = applyParams(
      {
        connection: {
          encrypted: true
        },
        headers: {
          host: 'example.com'
        },
        url: '/posts/1/comments/2?sort=date&filter=false'
      },
      {
        path: '/posts/:id/comments/:commentId',
        handler: jest.fn(),
        matchResult: ['', 1, 2],
        keys: [{ name: 'id' }, { name: 'commentId' }]
      })

    expect(req).toMatchSnapshot()
  })
})
