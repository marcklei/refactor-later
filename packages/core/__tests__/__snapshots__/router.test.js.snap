// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`applyParams ... applys the params and query to the req 1`] = `
Object {
  "connection": Object {
    "encrypted": true,
  },
  "headers": Object {
    "host": "example.com",
  },
  "params": Object {
    "commentId": 2,
    "id": 1,
  },
  "query": Object {
    "filter": "false",
    "sort": "date",
  },
  "url": "/posts/1/comments/2?sort=date&filter=false",
}
`;

exports[`mergeRoutes ... returns a merged routes object 1`] = `
Object {
  "ACL": Array [],
  "BIND": Array [],
  "CHECKOUT": Array [],
  "CONNECT": Array [],
  "COPY": Array [],
  "DELETE": Array [],
  "GET": Array [
    Object {
      "handler": [Function],
      "path": "/",
    },
    Object {
      "handler": [Function],
      "path": "/testroute",
    },
  ],
  "HEAD": Array [],
  "LINK": Array [],
  "LOCK": Array [],
  "M-SEARCH": Array [],
  "MERGE": Array [],
  "MKACTIVITY": Array [],
  "MKCALENDAR": Array [],
  "MKCOL": Array [],
  "MOVE": Array [],
  "NOTIFY": Array [],
  "OPTIONS": Array [],
  "PATCH": Array [],
  "POST": Array [
    Object {
      "handler": [Function],
      "path": "/group/:id",
    },
    Object {
      "handler": [Function],
      "path": "/user/:id",
    },
  ],
  "PROPFIND": Array [],
  "PROPPATCH": Array [],
  "PURGE": Array [],
  "PUT": Array [],
  "REBIND": Array [],
  "REPORT": Array [],
  "SEARCH": Array [],
  "SOURCE": Array [],
  "SUBSCRIBE": Array [],
  "TRACE": Array [],
  "UNBIND": Array [],
  "UNLINK": Array [],
  "UNLOCK": Array [],
  "UNSUBSCRIBE": Array [],
}
`;
